scalding-core/src/main/scala/com/twitter/scalding/typed/Grouped.scala

```diff
@@ -104,8 +104,27 @@ object Grouped {
       val keyF = new Fields("key")
       keyF.setComparator("key", new CascadingBinaryComparator(boxordSer))
       val pipe = op(ts, keyF)
+
+      @annotation.tailrec
+      def go(p: Pipe, visited: Set[Pipe], toVisit: List[Pipe]): Set[Pipe] = {
+        val notSeen: Set[Pipe] = p.getPrevious.filter(i => !visited.contains(i)).toSet
+        val nextVisited: Set[Pipe] = visited + p
+        val nextToVisit: List[Pipe] = (toVisit ++ notSeen).distinct
```

Comment:
this will hurt complexity if we get too large. Might be better to keep a `Set[Pipe]` too:

```scala
import scala.collection.immutable.Queue

case class ToVisit[T](queue: Queue[T], inQueue: Set[T]) {
  def maybeAdd(t: T): ToVisit[T] = if (inQueue(t)) this else {
   ToVisit(queue :+ t, inQueue + t)
  }
  def next: Option[(T, ToVisit[T])] =
    if (inQueue.isEmpty) None
    else Some(queue.head, ToVisit(queue.tail, toVisit - queue.head))
}
```